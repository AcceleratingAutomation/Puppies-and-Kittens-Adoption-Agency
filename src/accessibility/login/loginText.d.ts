export const usernameRequired: string;
export const passwordRequired: string;
export const usernameText: string;
export const passwordText: string;
export const loginText: string;
export const atLeast8Characters: string;
export const atLeastOneLowercaseLetter: string;
export const atLeastOneUppercaseLetter: string;
export const atLeastOneNumber: string;
export const atLeastOneSpecialCharacter: string;
export const errorLoggingIntoApp: string;
