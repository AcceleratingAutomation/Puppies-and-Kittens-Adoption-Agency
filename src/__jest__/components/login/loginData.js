import {
  required,
  atLeast8Characters,
  atLeastOneLowercaseLetter,
  atLeastOneUppercaseLetter,
  atLeastOneNumber,
  atLeastOneSpecialCharacter,
  errorLoggingIntoApp,
} from "../../../accessibility/login/loginText";

// Login Data Scenarios
// testTitle, username, password
// The username and password are just a mocked login. Do NOT add real passwords here.
export const validLoginScenarios = [
  ["valid credentials", "username", "Password123!"],
];

// testTitle, username, password, expectedErrorMessages
// Mocking an incorrect username or password combination. The form should be submitted but the user should not get logged in.
export const noLoginScenarios = [
  [
    "invalid credentials",
    "invalid username",
    "Invalid1$-password",
    errorLoggingIntoApp,
  ],
];

// testTitle, username, password, expectedErrorMessages
// These scenarios should not have a form submission because something is invalid.
export const noLoginOrSubmitScenarios = [
  ["empty username and password", "", "", required, required],
  ["empty username", "", "Password123!", required],
  ["empty password", "username", "", required],
  ["password too short", "username", "1234567", atLeast8Characters],
  [
    "password without uppercase letter",
    "username",
    "password123!",
    atLeastOneUppercaseLetter,
  ],
  [
    "password without lowercase letter",
    "username",
    "PASSWORD123!",
    atLeastOneLowercaseLetter,
  ],
  ["password without number", "username", "Password!", atLeastOneNumber],
  [
    "password without special character",
    "username",
    "Password123",
    atLeastOneSpecialCharacter,
  ],
  [
    "empty username and password too short",
    "",
    "1234567",
    required,
    atLeast8Characters,
  ],
  [
    "empty username and password without uppercase letter",
    "",
    "password123!",
    required,
    atLeastOneUppercaseLetter,
  ],
  [
    "empty username and password without lowercase letter",
    "",
    "PASSWORD123!",
    required,
    atLeastOneLowercaseLetter,
  ],
  [
    "empty username and password without number",
    "",
    "Password!",
    required,
    atLeastOneNumber,
  ],
  [
    "empty username and password without special character",
    "",
    "Password123",
    required,
    atLeastOneSpecialCharacter,
  ],
];
