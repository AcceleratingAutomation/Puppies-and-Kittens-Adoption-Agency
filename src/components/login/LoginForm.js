import React from 'react';
import { Grid, Typography, TextField, Button, Hidden } from "@material-ui/core";
import Tooltip from '@material-ui/core/Tooltip';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import { Field, Form, ErrorMessage } from 'formik';
import RescueImage from "../rescues/RescueImage";

const LoginForm = ({ isSubmitting }) => (
    <Form>
        <Grid
            container
            direction={"column"}
            alignItems={"center"}
            style={{ marginTop: "3vh" }}
        >
            <Grid item style={{ marginBottom: "3vh" }}>
                <Typography variant={"h3"}>
                    Puppies and Kittens <Hidden xsDown><br /></Hidden> Adoption Agency!
                </Typography>
            </Grid>
            <Grid item style={{ marginBottom: "3vh" }}>
                <Field
                    name="username"
                    as={TextField}
                    id={"username-input"}
                    label={"Username"}
                />
                <ErrorMessage name="username" component="div" style={{ color: 'red' }} />
            </Grid>
            <Grid item style={{ marginBottom: "5vh" }}>
                <Field
                    name="password"
                    as={TextField}
                    id={"password-input"}
                    label={
                        <React.Fragment>
                            Password
                            <Tooltip title={
                                <React.Fragment>
                                    <Typography variant="body2" style={{ fontSize: '1.2em' }}>
                                        Must contain:
                                        <ul>
                                            <li>at least 8 characters,</li>
                                            <li>1 lowercase letter,</li>
                                            <li>1 uppercase letter, and</li>
                                            <li>1 number</li>
                                        </ul>
                                    </Typography>
                                </React.Fragment>
                            }>
                                <HelpOutlineIcon fontSize="small" style={{ marginLeft: '10px' }} />
                            </Tooltip>
                        </React.Fragment>
                    }
                    type={"password"}
                />
                <ErrorMessage name="password" component="div" style={{ color: 'red' }} />
            </Grid>
            <Grid item style={{ marginBottom: "2vh" }}>
                <Button
                    style={{ margin: '0.625rem' }}
                    aria-label={"login"}
                    variant={"contained"}
                    size={"large"}
                    color={"primary"}
                    type="submit"
                    disabled={isSubmitting}
                >
                    LOGIN
                </Button>
            </Grid>
            <ErrorMessage name="login" component="div" style={{ color: 'red' }} />
            <Grid container style={{ margin: '0 auto', maxWidth: '80%' }} justify="center">
                <Grid item xs={12} sm={4}>
                    <RescueImage
                        type={"dog"}
                        image={0}
                        name={"Playful Puppy"}
                        width='15rem'
                        height='15rem'
                    />
                </Grid>
                <Grid item xs={12} sm={4}>
                    <RescueImage
                        type={"cat"}
                        image={0}
                        name={"Playful Kitten"}
                        width='15rem'
                        height='15rem'
                    />
                </Grid>
            </Grid>
        </Grid>
    </Form>
);

export default LoginForm;