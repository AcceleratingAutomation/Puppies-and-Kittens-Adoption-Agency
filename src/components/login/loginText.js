const usernameText = "Username";
// Allowing exception in gitleaks as this is a reference to the word Password
const passwordText = "Password";
const loginText = "LOGIN";
const required = "Required";
const atLeast8Characters = "Password must be at least 8 characters.";
const atLeastOneLowercaseLetter =
  "Password must contain at least one lowercase letter.";
const atLeastOneUppercaseLetter =
  "Password must contain at least one uppercase letter.";
const atLeastOneNumber = "Password must contain at least one number.";
const atLeastOneSpecialCharacter =
  "Password must contain at least one special character.";
const errorLoggingIntoApp = "Username or password is incorrect.";

module.exports = {
  required,
  usernameText,
  passwordText,
  loginText,
  atLeast8Characters,
  atLeastOneLowercaseLetter,
  atLeastOneUppercaseLetter,
  atLeastOneNumber,
  atLeastOneSpecialCharacter,
  errorLoggingIntoApp,
};
