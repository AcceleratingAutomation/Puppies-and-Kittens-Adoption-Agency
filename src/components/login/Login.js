import React, { useEffect, useRef } from "react";
import { useHistory } from "react-router-dom";
import { updateAppSettings } from "../../utils";
import { Formik } from 'formik';
import * as Yup from 'yup';
import { login } from '../../server/api/authApi';
import LoginForm from './LoginForm';

const LoginSchema = Yup.object().shape({
  username: Yup.string().required('Required'),
  password: Yup.string()
    .required('Required')
    .min(8, 'Password must be at least 8 characters')
    .matches(/[a-z]/, 'Password must contain at least one lowercase letter')
    .matches(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .matches(/[0-9]/, 'Password must contain at least one number'),
  login: Yup.string(),
});

export const Login = () => {
  const history = useHistory();
  const isMounted = useRef(false);

  useEffect(() => {
    isMounted.current = true;
    return () => {
      isMounted.current = false;
    };
  }, []);

  const onSubmit = async (values, { setSubmitting, setErrors }) => {
    try {
      const json = await login(values.username, values.password);
      if (isMounted.current) {
        if (json.message) {
          setErrors({ login: json.message });
        }
        else {
          updateAppSettings(json.token);
          history.push("/v1/rescues");
        }
        if (isMounted.current) {
          setSubmitting(false);
        }
      }
    } catch (err) {
      if (isMounted.current) {
        console.log("Error logging into app ", err.message);
      }
    }
  };

  return (
    <Formik
      initialValues={{ username: '', password: '', login: '' }}
      validationSchema={LoginSchema}
      onSubmit={onSubmit}
    >
      {({ isSubmitting }) => (
        <LoginForm isSubmitting={isSubmitting} />
      )}
    </Formik>
  );
};