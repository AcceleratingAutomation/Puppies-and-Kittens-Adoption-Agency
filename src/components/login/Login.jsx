import React, { useEffect, useRef } from "react";
import { useHistory } from "react-router-dom";
import { Formik } from "formik";
import * as Yup from "yup";
import { updateAppSettings } from "../../utils/utils";
import { login } from "../../server/apiService/authApi";
import LoginForm from "./LoginForm";
import {
  required,
  atLeast8Characters,
  atLeastOneLowercaseLetter,
  atLeastOneUppercaseLetter,
  atLeastOneNumber,
  atLeastOneSpecialCharacter,
} from "./loginText";

const LoginSchema = Yup.object().shape({
  username: Yup.string().required(required),
  password: Yup.string()
    .required(required)
    .min(8, atLeast8Characters)
    .matches(/[a-z]/, atLeastOneLowercaseLetter)
    .matches(/[A-Z]/, atLeastOneUppercaseLetter)
    .matches(/[0-9]/, atLeastOneNumber)
    .matches(/[\^$*.[\]{}()?\-"!@#%&/,><':;|_~`]/, atLeastOneSpecialCharacter),
  login: Yup.string(),
});

export default function Login() {
  const history = useHistory();
  const isMounted = useRef(false);

  useEffect(() => {
    isMounted.current = true;
    return () => {
      isMounted.current = false;
    };
  }, []);

  const onSubmit = async (values, { setSubmitting, setErrors }) => {
    try {
      const json = await login(values.username, values.password);
      if (isMounted.current) {
        if (json.message) {
          setErrors({ login: json.message });
        } else {
          updateAppSettings(json.token);
          history.push("/v1/rescues");
        }
        if (isMounted.current) {
          setSubmitting(false);
        }
      }
    } catch (err) {
      if (isMounted.current) {
        throw new Error(`Error logging into app ${err.message}`);
      }
    }
  };

  return (
    <Formik
      initialValues={{ username: "", password: "", login: "" }}
      validationSchema={LoginSchema}
      onSubmit={onSubmit}
    >
      {({ isSubmitting }) => <LoginForm isSubmitting={isSubmitting} />}
    </Formik>
  );
}
