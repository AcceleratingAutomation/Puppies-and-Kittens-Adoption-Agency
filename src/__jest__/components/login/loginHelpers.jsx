import React from "react";
import { Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import Login from "../../../components/login/Login";
import { login } from "../../../server/apiService/authApi";
import {
  loginText,
  passwordText,
  usernameText,
} from "../../../accessibility/login/loginText";
import { rescuesEndpoint } from "../../../server/apiService/apiConfig";

/**
 * Sets up the test environment for the Login component.
 * @param {Function} mockLogin - The mock login function.
 * @returns {Object} The utilities from @testing-library/react's render function, the mock login function, and the mock history push function.
 */
export const setup = (mockLogin) => {
  login.mockImplementation(mockLogin);

  const history = createMemoryHistory();
  const mockHistoryPush = jest.spyOn(history, "push");

  const utils = render(
    <Router history={history}>
      <Login />
    </Router>,
  );

  return { ...utils, mockLogin, mockHistoryPush };
};

/**
 * Fills the login form with the provided username and password and clicks the login button.
 * @param {string} username - The username to input.
 * @param {string} password - The password to input.
 * @param {Function} getByLabelText - The getByLabelText function from @testing-library/react.
 * @param {Function} getByText - The getByText function from @testing-library/react.
 */
export const fillForm = (username, password, getByLabelText, getByText) => {
  userEvent.type(getByLabelText(usernameText), username);
  userEvent.type(getByLabelText(passwordText), password);
  userEvent.click(getByText(loginText));
};

/**
 * Checks that the mock login function was called with the provided valid username and password.
 * Also, checks that after successful validation the mock history push function was called.
 * @param {string} username - The expected username.
 * @param {string} password - The expected password.
 * @param {Function} mockLogin - The mock login function.
 * @param {Function} mockHistoryPush - The mock history push function.
 */
export const checkSuccessfulLoginAndRedirect = async (
  username,
  password,
  mockLogin,
  mockHistoryPush,
) => {
  await act(async () => {
    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith(username, password);
      expect(mockHistoryPush).toHaveBeenCalledWith(rescuesEndpoint);
    });
  });
};

/**
 * Checks that the mock login function was called with the provided valid username and password.
 * Also, checks that the mock history push function was not called when an error occurred.
 * @param {string} username - The expected username.
 * @param {string} password - The expected password.
 * @param {Function} mockLogin - The mock login function.
 * @param {Function} mockHistoryPush - The mock history push function.
 */
export const checkLoginCalledButNoRedirect = async (
  username,
  password,
  mockLogin,
  mockHistoryPush,
) => {
  await act(async () => {
    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith(username, password);
      expect(mockHistoryPush).not.toHaveBeenCalled();
    });
  });
};

/**
 * Checks that the mock login function and the mock history push function were not called with incomplete or incorrect form data.
 * @param {Function} mockLogin - The mock login function.
 * @param {Function} mockHistoryPush - The mock history push function.
 */
export const checkNoLoginOrRedirect = async (mockLogin, mockHistoryPush) => {
  await act(async () => {
    await waitFor(() => {
      expect(mockLogin).not.toHaveBeenCalled();
      expect(mockHistoryPush).not.toHaveBeenCalled();
    });
  });
};

/**
 * Checks that the expected error messages are displayed in the DOM.
 * @param {Function} findByText - The findByText function from @testing-library/react.
 * @param {string} error1 - The first error message to check.
 * @param {string} [error2=""] - The second error message to check. Defaults to an empty string because several scenarios only display one error.
 * Note: Ran into React console errors, for scenarios that display multiple errors, when using an array of error messages instead of the two variables.
 */
export const checkErrorMessages = async (findByText, error1, error2 = "") => {
  let errorMessage = "";
  if (error1) {
    errorMessage = await findByText(error1);
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toMatchSnapshot();
  }

  if (error2) {
    errorMessage = await findByText(error2);
    expect(errorMessage).toBeInTheDocument();
    expect(errorMessage).toMatchSnapshot();
  }
};
