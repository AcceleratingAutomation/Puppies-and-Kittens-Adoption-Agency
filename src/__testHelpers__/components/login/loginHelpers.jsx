import React from "react";
import { Router } from "react-router-dom";
import { render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import Login from "../../../components/login/Login";
import { login } from "../../../server/apiService/authApi";
import {
  loginText,
  passwordText,
  usernameText,
} from "../../../accessibility/login/loginText";

/**
 * Sets up the test environment for the Login component.
 * @param {Function} mockLogin - The mock login function.
 * @returns {Object} The utilities from @testing-library/react's render function, the mock login function, and the mock history push function.
 */
export const setup = (mockLogin) => {
  login.mockImplementation(mockLogin);

  const history = createMemoryHistory();
  const mockHistoryPush = jest.spyOn(history, "push");

  const utils = render(
    <Router history={history}>
      <Login />
    </Router>,
  );

  return { ...utils, mockLogin, mockHistoryPush };
};

/**
 * Fills the login form with the provided username and password and clicks the login button.
 * @param {string} username - The username to input.
 * @param {string} password - The password to input.
 * @param {Function} getByLabelText - The getByLabelText function from @testing-library/react.
 * @param {Function} getByText - The getByText function from @testing-library/react.
 */
export const fillForm = (username, password, getByLabelText, getByText) => {
  userEvent.type(getByLabelText(usernameText), username);
  userEvent.type(getByLabelText(passwordText), password);
  userEvent.click(getByText(loginText));
};

/**
 * Checks that the mock login function was called with the provided valid username and password.
 * Also, that the mock history push function was called with "/v1/rescues".
 * @param {string} username - The expected username.
 * @param {string} password - The expected password.
 * @param {Function} mockLogin - The mock login function.
 * @param {Function} mockHistoryPush - The mock history push function.
 */
export const checkSuccessfulLoginAndRedirect = async (
  username,
  password,
  mockLogin,
  mockHistoryPush,
) => {
  await act(async () => {
    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith(username, password);
      expect(mockHistoryPush).toHaveBeenCalledWith("/v1/rescues");
    });
  });
};

/**
 * Checks that the mock login function and the mock history push function were not called with incomplete or incorrect form data.
 * @param {Function} mockLogin - The mock login function.
 * @param {Function} mockHistoryPush - The mock history push function.
 */
export const checkNoLoginOrRedirect = async (mockLogin, mockHistoryPush) => {
  await act(async () => {
    await waitFor(() => {
      expect(mockLogin).not.toHaveBeenCalled();
      expect(mockHistoryPush).not.toHaveBeenCalled();
    });
  });
};

/**
 * Checks that the expected error messages are displayed.
 * @param {Function} findAllByText - The findAllByText function from @testing-library/react. Using findAllByText because 'Required' can show up twice.
 * @param {string} text1 - The first error message to check.
 * @param {string} [text2=""] - The second error message to check. Defaults to an empty string.
 */
export const checkErrorMessages = async (findAllByText, text1, text2 = "") => {
  if (text1) {
    const errorMessages1 = await findAllByText(text1);
    errorMessages1.forEach((errorMessage) => {
      expect(errorMessage).toBeInTheDocument();
    });
  }

  if (text2) {
    const errorMessages2 = await findAllByText(text2);
    errorMessages2.forEach((errorMessage) => {
      expect(errorMessage).toBeInTheDocument();
    });
  }
};
