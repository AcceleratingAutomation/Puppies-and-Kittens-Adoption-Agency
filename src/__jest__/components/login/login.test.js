import {
  checkErrorMessages,
  checkLoginCalledButNoRedirect,
  checkNoLoginOrRedirect,
  checkSuccessfulLoginAndRedirect,
  fillForm,
  setup,
} from "./loginHelpers";
import noLoginOrSubmitScenarios from "./loginData";
import { errorLoggingIntoApp } from "../../../accessibility/login/loginText";

// Mock the login function from the authApi module
jest.mock("../../../server/apiService/authApi", () => ({
  login: jest.fn(),
}));

// Mock the updateAppSettings function from the utils module
jest.mock("../../../utils/utils", () => ({
  updateAppSettings: jest.fn(),
}));

/**
 * This test suite tests the functionality of the Login component.
 */
describe("Login component", () => {
  let getByLabelText;
  let getByText;
  let findByText;
  let mockLogin;
  let mockHistoryPush;
  let asFragment;

  describe("", () => {
    beforeEach(() => {
      const mockLoginFunction = jest
        .fn()
        .mockResolvedValue({ token: "fakeToken" });
      ({ getByLabelText, getByText, findByText, mockLogin, mockHistoryPush } =
        setup(mockLoginFunction));
    });

    describe("should login with", () => {
      it("valid credentials", async () => {
        // The username and password are just a mocked login. Do NOT add real passwords here.
        const username = "username";
        const password = "Password123!";
        fillForm(username, password, getByLabelText, getByText);
        await checkSuccessfulLoginAndRedirect(
          username,
          password,
          mockLogin,
          mockHistoryPush,
        );
      });
    });

    describe.each(noLoginOrSubmitScenarios)(
      "should not login or submit form with",
      (testTitle, username, password, expectedError1, expectedError2) => {
        it(testTitle, async () => {
          fillForm(username, password, getByLabelText, getByText);
          await checkErrorMessages(findByText, expectedError1, expectedError2);
          await checkNoLoginOrRedirect(mockLogin, mockHistoryPush);
        });
      },
    );
  });

  describe("should submit the form but not login", () => {
    it("when a general error message is displayed", async () => {
      const username = "username";
      const password = "Password1!";
      const errorMessage = "Error logging into app.";

      const mockLoginFunction = jest
        .fn()
        .mockRejectedValue(new Error(errorMessage));
      ({ getByLabelText, getByText, findByText, mockLogin, mockHistoryPush } =
        setup(mockLoginFunction));

      fillForm(username, password, getByLabelText, getByText);
      await checkErrorMessages(findByText, errorMessage);
      await checkLoginCalledButNoRedirect(
        username,
        password,
        mockLogin,
        mockHistoryPush,
      );
    });

    it("with invalid credential", async () => {
      const username = "invalid_username";
      const password = "Invalid1$-password";

      const mockLoginFunction = jest
        .fn()
        .mockResolvedValue({ message: errorLoggingIntoApp });
      ({ getByLabelText, getByText, findByText, mockLogin, mockHistoryPush } =
        setup(mockLoginFunction));

      fillForm(username, password, getByLabelText, getByText);
      await checkErrorMessages(findByText, errorLoggingIntoApp);
      await checkLoginCalledButNoRedirect(
        username,
        password,
        mockLogin,
        mockHistoryPush,
      );
    });
  });

  describe("should render correctly", () => {
    it("in snapshot comparison after setup", () => {
      const { asFragment: fragment } = setup(mockLogin);
      asFragment = fragment;
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
