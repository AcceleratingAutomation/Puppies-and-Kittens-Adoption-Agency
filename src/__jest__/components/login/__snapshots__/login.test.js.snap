// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component  should not login or submit form with empty password 1`] = `
<div
  style="color: red;"
>
  Password Required
</div>
`;

exports[`Login component  should not login or submit form with empty username 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password 2`] = `
<div
  style="color: red;"
>
  Password Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password too short 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password too short 2`] = `
<div
  style="color: red;"
>
  Password must be at least 8 characters.
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without lowercase letter 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without lowercase letter 2`] = `
<div
  style="color: red;"
>
  Password must contain at least one lowercase letter.
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without number 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without number 2`] = `
<div
  style="color: red;"
>
  Password must contain at least one number.
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without special character 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without special character 2`] = `
<div
  style="color: red;"
>
  Password must contain at least one special character.
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without uppercase letter 1`] = `
<div
  style="color: red;"
>
  Username Required
</div>
`;

exports[`Login component  should not login or submit form with empty username and password without uppercase letter 2`] = `
<div
  style="color: red;"
>
  Password must contain at least one uppercase letter.
</div>
`;

exports[`Login component  should not login or submit form with password too short 1`] = `
<div
  style="color: red;"
>
  Password must be at least 8 characters.
</div>
`;

exports[`Login component  should not login or submit form with password without lowercase letter 1`] = `
<div
  style="color: red;"
>
  Password must contain at least one lowercase letter.
</div>
`;

exports[`Login component  should not login or submit form with password without number 1`] = `
<div
  style="color: red;"
>
  Password must contain at least one number.
</div>
`;

exports[`Login component  should not login or submit form with password without special character 1`] = `
<div
  style="color: red;"
>
  Password must contain at least one special character.
</div>
`;

exports[`Login component  should not login or submit form with password without uppercase letter 1`] = `
<div
  style="color: red;"
>
  Password must contain at least one uppercase letter.
</div>
`;

exports[`Login component should render correctly in snapshot comparison after setup 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
      style="margin-top: 3vh;"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="margin-bottom: 3vh;"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
        >
          Puppies and Kittens  Rescue and Adoption Agency!
        </h3>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="margin-bottom: 3vh;"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="username-input"
            id="username-input-label"
          >
            Username
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username-input"
              name="username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Username
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="margin-bottom: 5vh;"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password-input"
            id="password-input-label"
          >
            Password
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="HelpOutlineIcon"
              focusable="false"
              style="margin-left: 10px;"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 18h2v-2h-2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4"
              />
            </svg>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password-input"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Password
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                    data-mui-internal-clone-element="true"
                    data-testid="HelpOutlineIcon"
                    focusable="false"
                    style="margin-left: 10px;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 18h2v-2h-2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4"
                    />
                  </svg>
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        style="margin-bottom: 2vh;"
      >
        <button
          aria-label="login"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary css-jh47zj-MuiButtonBase-root-MuiButton-root"
          style="margin: 0.625rem;"
          tabindex="0"
          type="submit"
        >
          LOGIN
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-1lym95h-MuiGrid-root"
        style="margin: 0px auto; max-width: 80%;"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wg4xta-MuiGrid-root"
        >
          <img
            alt="Playful Puppy's picture"
            class="responsive-image"
            data-testid="display-image"
            src="/images/dog/placeholder-puppy.webp"
            style="border-radius: 25%; width: 15rem; height: 15rem; object-fit: cover;"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-wg4xta-MuiGrid-root"
        >
          <img
            alt="Playful Kitten's picture"
            class="responsive-image"
            data-testid="display-image"
            src="/images/cat/placeholder-kitten.webp"
            style="border-radius: 25%; width: 15rem; height: 15rem; object-fit: cover;"
          />
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Login component should submit the form but not login when a general error message is displayed 1`] = `
<div
  style="color: red;"
>
  Error logging into app: Failed to fetch
</div>
`;

exports[`Login component should submit the form but not login with invalid credential 1`] = `
<div
  style="color: red;"
>
  Username or password is incorrect.
</div>
`;
